fairworkflows/fairstep.py

```diff
@@ -66,24 +68,52 @@ def load_from_nanopub(self, uri):
             if len(concepts_introduced) == 0:
                 raise ValueError('This nanopub does not introduce any concepts. Please provide URI to the step itself (not just the nanopub).')
             elif len(concepts_introduced) > 0:
-                step_uri = concepts_introduced[0]
+                step_uri = str(concepts_introduced[0])
 
             print('Assuming step URI is', step_uri)
 
         else:
             step_uri = uri
 
         self._uri = step_uri
-        self.self_ref = rdflib.URIRef(self._uri)
 
         # Check that the nanopub's assertion actually contains triples refering to the given step's uri
-        if (rdflib.URIRef(self.self_ref), None, None) not in np.assertion:
+        if (rdflib.URIRef(self._uri), None, None) not in np.assertion:
             raise ValueError(f'No triples pertaining to the specified step (uri={step_uri}) were found in the assertion graph of the corresponding nanopublication (uri={np_uri})')
 
         # Else extract all triples in the assertion into the rdf graph for this step
         self._rdf = rdflib.Graph()
         self._rdf += np.assertion
 
+        # Record that this RDF originates from a published source
+        self._is_published = True
+
+
+    def publish_as_nanopub(self):
+        """
+        Publishes the rdf for this FairStep as a nanopublication.
+        Returns True if published successfully.
+        """
+
+        # If this step has been modified from a previously publised step, include this in the derived_from PROV (if applicable)
+        derived_from = None
+        if self._is_published is True:
+            if self.is_modified is True:
+                derived_from = self._uri
+            else:
+                warnings.warn(f'Cannot publish() FairStep. This step is already published (at {self._uri}) and has not been modified.')
```

Comment:
I felt an error might be too strong, since something hasn't really gone wrong. The step is still published, with the correct URI, it's just that nothing has changed. Is throwing an exception preferable?

Reply:
Yeah, I guess you are right. This means a user would expect this behavior (i.e. not throwing an error), and does not really want to be informed on whether the workflow is already published or not. Which makes sense I guess.